docs/proposals/volume-provisioning.md

```diff
@@ -0,0 +1,315 @@
+
+1.  Storage classes should be able to hold a set of provisioner parameters
+2.  The provisioner should be passed the claim and the details of the storage
+    class it is provisioning for
+
+#### Out-of-tree provisioners
+
+One of our goals is to enable administrators to create out-of-tree
+provisioners, that is, provisioners whose code does not live in the Kubernetes
+project.  Our experience since the 1.2 release with dynamic provisioning has
+shown that it is impossible to anticipate every aspect and manner of
+provisioning that administrators will want to perform.  The proposed design
+should not prevent future work to allow out-of-tree provisioners.
+
+## Analysis
+
+### Summary of alpha functionality
+
+The alpha functionality for dynamic provisioning works as follows:
+
+1.  A cluster has only a single provisioner at a time, dictated by the cloud
+    provider
+2.  When a new claim is submitted, the controller attempts to find an existing
+    volume that will fulfill the claim.  If a suitable volume is found, the
+    controller binds the claim and volume together and carries on normally.
+3.  If no volume is found for the claim, and the claim has the annotation
+    `volume.alpha.kubernetes.io/storage-class`, the provisioner is invoked
+    to provision a volume to fill that claim
+4.  All provisioners are in-tree; they implement an interface called
+    `ProvisionableVolumePlugin` which has a method, `NewProvisioner()`,
+    that returns a new `Provisioner` instance
+5.  The controller calls the provisioner's `Provision()` method; `Provision`
+    is responsible for provisioning the volume in the cloud provider and
+    returns an `api.PersistentVolume` and an error
+6.  If `Provision` returns an error, the controller stops trying to provision
+7.  If `Provision` returns no error, the controller creates the returned
+    `api.PersistentVolume`, already bound to the claim
+  1.  If the create operation fails, it is retried
+  2.  If the create operation never succeeds, the controller attempts to delete
+      the provisioned volume and creates an event on the claim
+
+### When should we provision volumes?
+
+It is sensible to provision volumes only when an existing volume cannot
+satisfy a claim.  This ensures that existing, already allocated resources are
+used before additional resources are allocated.  In this regard, the existing
+logic need not change much when storage classes are added.  The primary
+difference is that in the existing design, there is no way to parameterize
+provisioners or have more than one provisioner per cluster.  Therefore, once
+we know that we have to provision a volume to satisfy a claim, we need to
+determine which storage class to use for the claim.
+
+### How should a claim be matched to a storage class?
+
+We've discussed so far that volume claims and storage classes will be matched
+using the claim's label selector.  With that said, there are a number of small
```

Comment:
Ultimately all PVC's need to be quota'd.  We *absolutely* want quota on both number of volumes and total capacity of volumes per class.  It's not totally clear to me whether we need a deeper quota on individual volumes per zone, but I think it'll be likely that your quota allows you access to 0 volumes in zone "east" and 5 volumes in zone "west".  I think quota on labels makes more sense than quota on storage class, because having to create individual classes for zones is hard to manage and requires pointless duplication.

I prefer simple storage classes for most user experience points of view, but if we have to have two mechanisms to quota access to storage that starts to sound wrong to me.